### DISCUSSION TITLE
Just the raw base64?

### DISCUSSION THREAD

**Discussion by strider72** [chillerlan/php-qrcode#185] [General]
Hi! I'm trying to get this to work with a 2FA library. The format for the library is to get two separate values: 1) mimetype (e.g. `image/svg+xml`), and 2) the image itself in base64.

Your plugin pre-combines these values in the string when it renders as dataurl.

Is there any way to easily get the base64 image string?  Am I best off just exploding the string at the comma and grabbing the second element?

**Comment by codemasher** [comment]
Hey! The base64 data URI is a convenience functionality so that the output can be used directly in an `<img src="" />` tag without further processing (e.g. cache a .png file). So yes, the simplest way is to just explode the data URI on the comma `,`. Alternatively, you could set `QROptions::$imageBase64` to `false` and then `base64_encode()` the content by yourself, but i guess that's more overhead.

**Comment by strider72** [comment]
Is that actually more overhead? I'm just moving the encode from one place to another!

Thanks, I'll give it a shot.

**Comment by strider72** [comment]
Got it working. For anyone else who's looking, here is my working code for user Chillerlan QRs with RobThree's TOTP library:
https://robthree.github.io/TwoFactorAuth/

```php
<?php
namespace [redacted];
use chillerlan\QRCode as qr;
use chillerlan\Settings\SettingsContainerInterface;
use RobThree\Auth\Providers\Qr as mfa;

class ChillerlanQRCodeProvider extends qr\QRCode implements mfa\IQRCodeProvider {

	function __construct( SettingsContainerInterface $options = NULL ) {
		$options ??= new qr\QROptions( [
			'outputType' => qr\QRCode::OUTPUT_MARKUP_SVG,
			'eccLevel' => qr\QRCode::ECC_M,
		] );
		$options->imageBase64 = False;	// imageBase64 above must always be False. Override if necessary.
		parent::__construct( $options );
	}

	public function getQRCodeImage( string $qrtext, int $size ): string {
		return self::render( $qrtext );
	}

	public function getMimeType(): string {
		return match( $this->options->outputType ) {
			self::OUTPUT_MARKUP_SVG=>'image/svg+xml',
			self::OUTPUT_IMAGE_PNG=>'image/png',
			self::OUTPUT_IMAGE_JPG=>'image/jpeg',
			self::OUTPUT_IMAGE_GIF=>'image/gif',
		};
	}
}
```

In your main code, use it when creating your 2FA object. (Named param syntax is PHP >=8.0):

```php
$mfa = new TwoFactorAuth(
  issuer: 'My Company Name',
  qrcodeprovider: new ChillerlanQRCodeProvider()
);
```

NOTE: In the above class, I have set options that are my subjective choices.  In your usage, `__construct` could be as simple as...

```php
function __construct( SettingsContainerInterface $options = NULL ) {
	$options ??= new qr\QROptions();
	$options->imageBase64 = False;	// imageBase64 above must always be False. Override if necessary.
	parent::__construct( $options );
}
```

**Reply by codemasher** [reply]
Nice! Which btw reminds me to update [my own MFA library](https://github.com/chillerlan/php-authenticator).... hehe

re the `FIXME` - why don't you just add a line `$options->imageBase64 = false;` before you hand the options instance into the parent constructor?

**Reply by strider72** [reply]
Fixed. Looking at the code for `QROptionsTrait` I didn't think I could set it after creation because there isn't an explicit set_ function for imageBase64

**Reply by codemasher** [reply]
*Freddie Mercury voice* "[It's a kind of magic!](https://github.com/chillerlan/php-settings-container/blob/91904fe64876ed75ab70bc0d43198638432528b8/src/SettingsContainerAbstract.php)" 🎶

**Reply by strider72** [reply]
Yeah, that's the code I looked at -- I just misread it.  I thought it only set it if the set_ function exists. In reality it sets it anyways, but uses the set_ function if it exists.

**Reply by codemasher** [reply]
I think i need to include the documentation for the settings container or at least place a link somewhere as it is not that obvious as i hoped. :D

**Reply by strider72** [reply]
Edited code to add a note about the `__construct` function. Examples now show with or without custom options.

**Reply by codemasher** [reply]
micro-optimization: `If( ! $options ) $options = new qr\QROptions();` to `$options ??= new qr\QROptions();` (this is exactly what the null coalesce assign was made for).

**Comment by strider72** [comment]
I would love better explanation of the QR customization (colors, icons, shapes). I chose your library for the customization, but have had trouble figuring it out.  Will have to take another look, as perhaps messing with this side-quest has shown me a few tricks I didn't realize I'd learned....)

**Reply by codemasher** [reply]
I'm close to the v5 finish line, once this is released, i'll take my time and edit the wiki accordingly. In the meantime, you can just open new discussions for things you want to know - this also helps me figuring out what needs better documentation.

**Reply by strider72** [reply]
I think the confusion comes from two things:

1. The break between v4 and v5. People are mixing and matching.  V5 release will fix this no doubt, as examples get updated.  For now I can't put "main" branch into a live site. 

2. In a lot of your examples you do many things at once.  You show shapes AND colors AND logos etc, and it's too hard to parse out how to do each separate thing.  Show us one thing at a time is far easier to understand.  Start with a plain QR and give three separate examples: color change; logo; dot shape.  Keep them totally separate. If we're reasonable competent we should be able to combine examples as needed. :-)

**Reply by codemasher** [reply]
Yep, the first point is on me. I could have let v4.x in the main branch for a bit longer because i should have known that my ADHD brain wouldn't pull through to get v5 done in a timely manner...

The second point is a result of helping other people asking for specific things in discussions and i agree that it is not necessarily helpful for understanding the basics.

**Reply by strider72** [reply]
Something that could be *really* useful is solid PHPdoc for each option in QROptionsTrait.  Thought I might use `$imageTransparencyBG` to make background transparent, but it doesn't seem to do anything.  It's probably not that complicated, but it's not explained -- just labelled.  Good PHPdoc info shows up in an IDE and is really easy to use.  (This may not help as much in all cases of course, especially across multiple classes.)

In my own code I'm becoming more and more strict about type hinting and similar, and it makes things easier for me later on when I don't remember what exactly I did when writing something.  Helps focus me so my code stays on track -- I'm setting rules for myself as I go.

**Reply by codemasher** [reply]
I absolutely agree with you here and i know i should add more information on several docblocks, but hey, i'm trying my best! (also, any help with docblocks is greatly appreciated!)

Since you mention `QROptionsTrait` specifically, and in case you're using an IDE other than PHPStorm: does the docblock linking work with protected magic properties in a trait? PS didn't even recognize these for a long time ([and still has problems](https://youtrack.jetbrains.com/issue/WI-66549)) and i had to add `@property` tags in `QROptions`.

**Reply by strider72** [reply]
I use phpStorm
